frontends/common/constantFolding.h

```diff
@@ -116,22 +116,24 @@ class DoConstantFolding : public Transform, public ResolutionContext {
  public:
     DoConstantFolding(const DeclarationLookup *refMap, const TypeMap *typeMap, bool warnings = true,
                       ConstantFoldingPolicy *policy = nullptr)
-        : refMap(refMap), typeMap(typeMap), typesKnown(typeMap != nullptr), warnings(warnings) {
-        if (policy) {
-            this->policy = policy;
-        } else {
-            this->policy = new ConstantFoldingPolicy();
-        }
+        : policy(policy ? policy : new ConstantFoldingPolicy()),
+          refMap(refMap),
+          typeMap(typeMap),
+          typesKnown(typeMap != nullptr),
+          warnings(warnings) {
         visitDagOnce = true;
         setName("DoConstantFolding");
         assignmentTarget = false;
     }
 
-    // If DeclarationLookup is not passed, then resolve by our own.
+    // If DeclarationLookup is not passed, then resolve by our own. We might
+    // need proper context for this
     explicit DoConstantFolding(const TypeMap *typeMap, bool warnings = true,
                                ConstantFoldingPolicy *policy = nullptr)
         : DoConstantFolding(this, typeMap, warnings, policy) {}
 
+    DoConstantFolding() : DoConstantFolding(nullptr, nullptr) {}
```

Comment:
Shouldn't this be `DoConstantFolding() : DoConstantFolding(this, nullptr) {}`

Reply:
There is one very subtle but important difference. We do use refmap in absence of typemap during constfolding of path expressions. However, this requires more or less complete IR. This is not the case for P14-to-16 conversion, where constfolding is done in the middle of conversion and therefore we just crash not being able to resolve anything.

On the other hand, it seems we're missing some constfolding opportunities if we'd require known types along with resolution contex.

Things are a bit messy, yes, I am planning to untangle this eventually.